test/cli_test.go

```diff
@@ -0,0 +1,379 @@
+	if err != nil {
+		t.Fatalf("marshaling: %v", err)
+	}
+	wantJSON, err := json.Marshal(want)
+	if err != nil {
+		t.Fatalf("marshaling: %v", err)
+	}
+	require.JSONEq(t, string(wantJSON), string(gotJSON))
+}
+
+func testValidateCommand(t *testing.T, wantError bool, want, dir, name string, offline bool, ruleName string) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("cannot get current directory: %v", err)
+	}
+	policyPath := filepath.Join(cwd, samplePolicyPath, ruleName)
+	var got []byte
+	switch version.LeastSupportedVersion() {
+	case version.TF11:
+		got = tfvValidate(t, wantError, dir, name+".tfplan", policyPath, offline)
+	default:
+		got = tfvValidate(t, wantError, dir, name+".tfplan.json", policyPath, offline)
+	}
+	wantRe := regexp.MustCompile(want)
+	if want != "" && !wantRe.Match(got) {
+		t.Fatalf("binary did not return expect output, \ngot=%s \nwant (regex)=%s", string(got), want)
+	}
+}
+
+func terraform(t *testing.T, dir, name string) {
+	switch version.LeastSupportedVersion() {
+	case version.TF11:
+		terraformInit(t, "terraform", dir)
+		terraformPlan(t, "terraform", dir, name+".tfplan")
+	default:
+		terraformInit(t, "terraform", dir)
+		terraformPlan(t, "terraform", dir, name+".tfplan")
+		payload := terraformShow(t, "terraform", dir, name+".tfplan")
+		saveFile(t, dir, name+".tfplan.json", payload)
+	}
+}
+
+func terraformInit(t *testing.T, executable, dir string) {
+	wantError := false
+	cmd := exec.Command(executable, "init", "-input=false", dir)
+	cmd.Env = []string{"HOME=" + filepath.Join(dir, "fakehome")}
+	cmd.Dir = dir
+	run(t, cmd, wantError)
+}
+
+func terraformPlan(t *testing.T, executable, dir, tfplan string) {
+	wantError := false
+	cmd := exec.Command(executable, "plan", "-input=false", "--out", tfplan, dir)
+	cmd.Env = []string{"HOME=" + filepath.Join(dir, "fakehome")}
+	cmd.Dir = dir
+	run(t, cmd, wantError)
+}
+
+func terraformShow(t *testing.T, executable, dir, tfplan string) []byte {
+	wantError := false
+	cmd := exec.Command(executable, "show", "--json", tfplan)
+	cmd.Env = []string{"HOME=" + filepath.Join(dir, "fakehome")}
+	cmd.Dir = dir
+	payload, _ := run(t, cmd, wantError)
+	return payload
+}
+
+func saveFile(t *testing.T, dir, filename string, payload []byte) {
+	fullpath := filepath.Join(dir, filename)
+	f, err := os.Create(fullpath)
+	if err != nil {
+		t.Fatalf("error while creating file %s, error %v", fullpath, err)
+	}
+	_, err = f.Write(payload)
+	if err != nil {
+		t.Fatalf("error while writing to file %s, error %v", fullpath, err)
+	}
+}
+
+func tfvConvert(t *testing.T, dir, tfplan string, offline bool) []byte {
+	executable := tfvBinary
+	wantError := false
+	args := []string{"convert", "--project", data.Provider["project"]}
+	if offline {
+		args = append(args, "--ancestry", data.Ancestry)
+	}
+	args = append(args, tfplan)
+	cmd := exec.Command(executable, args...)
+	cmd.Env = []string{"GOOGLE_APPLICATION_CREDENTIALS=" + data.Provider["credentials"]}
```

Comment:
If `TEST_CREDENTIALS` isn't explicitly set, then we're already relying on application default credentials and shouldn't override.

So let's make this line conditional on `data.Provider["credentials"]` not being nil.

Reply:
added.